core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,200 @@
+    // expire time is set to 6s
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled", "false")
+      .set("spark.scheduler.executorTaskBlacklistTime", "6000")
+
+    sc = new SparkContext(conf)
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf, clock)
+    // Executor 1 into blacklist at Time 00:00:00
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("1"))
+
+    clock.setTime(2000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("1"))
+    // Executor 1 failed again at Time 00::00:02
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, FAILURE)
+
+    clock.setTime(3000)
+    // Executor 2 failed at Time 00:00:03
+    tracker.updateFailureExecutors(stage1, taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("1", "2"))
+
+    clock.setTime(6000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("1", "2"))
+
+    clock.setTime(8000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("2"))
+
+    clock.setTime(10000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set())
+  }
+
+  test("blacklist feature is off by default") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled", "false")
+    sc = new SparkContext(conf)
+
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf, clock)
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, FAILURE)
+    tracker.updateFailureExecutors(stage1, taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set())
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set())
+
+    tracker.updateFailureExecutors(stage1, taskInfo_3_3_hostB, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 3) === Set())
+    assert(tracker.nodeBlacklist() === Set())
+  }
+
+  test("default strategy works") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled", "false")
+      .set("spark.scheduler.executorTaskBlacklistTime", "1000")
+    sc = new SparkContext(conf)
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf, clock)
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, FAILURE)
+    tracker.updateFailureExecutors(stage1, taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("1", "2"))
+    assert(tracker.executorBlacklist(scheduler, stage1, 2) === Set())
+
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, Success)
```

Comment:
can you add a really brief comment before each section (where you've put a blank line already) indicating what is going on?  eg, here: "task 1 succeeds on executor 1, so we remove it from the blacklist for executor 1", and for the next one, "task 3 succeeds on executor 3, no effect on blacklist for task 1", etc.